pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java

```
@@ -0,0 +1,272 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import org.apache.calcite.avatica.util.ByteString;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.NlsString;
+import org.apache.calcite.util.TimestampString;
+import org.apache.pinot.common.function.FunctionInfo;
+import org.apache.pinot.common.function.FunctionInvoker;
+import org.apache.pinot.common.function.FunctionRegistry;
+import org.apache.pinot.spi.utils.TimestampUtils;
+import org.apache.pinot.sql.parsers.SqlCompilationException;
+
+
+/**
+ * SingleValueAggregateRemoveRule that matches an Aggregate function SINGLE_VALUE and remove it
+ *
+ */
+public class PinotEvaluateLiteralRule extends RelOptRule {
+  public static final PinotEvaluateLiteralRule INSTANCE =
+      new PinotEvaluateLiteralRule(PinotRuleUtils.PINOT_REL_FACTORY);
+
+  public PinotEvaluateLiteralRule(RelBuilderFactory factory) {
+    super(operand(RelNode.class, any()), factory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    // Traverse the relational expression using a RexShuttle visitor
+    AtomicBoolean hasEvaluateCall = new AtomicBoolean(false);
+    call.rel(0).accept(new RexShuttle() {
+      @Override
+      public RexNode visitCall(RexCall call) {
```

Is there a way to early terminate? If no, at least we should check if `hasLiteralOnlyCall` is already set to true in the beginning

---

Did a little bit of research, and seems we can throw a custom exception to break the traverse, and based on whether the exception is thrown to identify if it has literal only call

---

No need to do that, just return at the first of the call is good enough.

---

Wouldn't this always traverse the whole tree? Notice that `RexShuttle` has all the default visit method that are not cheap (it might clone the node to check if it is changed)

---

Make sense, move to use `RelVisitor` for `onMatch`, it's just for visiting the nodes and set `hasLiteralOnlyCall`.